From bd14c4c050679861413c6a839b094a06c1b88273 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Thu, 29 Aug 2024 17:03:34 +0800
Subject: [PATCH 118/236] mtk: hostapd: update channel information in WBCS to
 carry maximal BW with no punctured sub-channel.

When AP is about to switch to a new channel with punctured sub-channel,
it will indicated the punctured bitmap in Bandwidth Indication
sub-element of Channel Switch Wrapper. However, the WBCS sub-element has
no field for punctured bitmap and is responsible for carring channel switch
information for legacy STAs, so the AP should calculate the maximal BW
with no punctured sub-channel and carry it in WBCS sub-element.

mtk: hostapd: do not indicate a bandwidth more than 320 MHz in WBCS element

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/ieee802_11.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 2d0d83b4d..9528c1a84 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -7782,6 +7782,10 @@ static u8 * hostapd_eid_wb_channel_switch(struct hostapd_data *hapd, u8 *eid,
 					  u8 chan1, u8 chan2)
 {
 	u8 bw;
+	enum oper_chan_width oper_chwidth = CONF_OPER_CHWIDTH_160MHZ;
+#ifdef CONFIG_IEEE80211BE
+	u16 punct_bitmap = hapd->cs_freq_params.punct_bitmap;
+#endif /* CONFIG_IEEE80211BE */
 
 	/* bandwidth: 0: 40, 1: 80, 160, 80+80, 4 to 255 reserved as per
 	 * IEEE Std 802.11-2024, 9.4.2.156 and Table 9-316 (VHT Operation
@@ -7802,6 +7806,7 @@ static u8 * hostapd_eid_wb_channel_switch(struct hostapd_data *hapd, u8 *eid,
 		else
 			chan1 += 16;
 
+		oper_chwidth = CONF_OPER_CHWIDTH_320MHZ;
 		/* fallthrough */
 	case 160:
 		/* Update the CCFS0 and CCFS1 values in the element based on
@@ -7824,15 +7829,25 @@ static u8 * hostapd_eid_wb_channel_switch(struct hostapd_data *hapd, u8 *eid,
 		break;
 	case 80:
 		bw = 1;
+		oper_chwidth = CONF_OPER_CHWIDTH_80MHZ;
 		break;
 	case 40:
 		bw = 0;
+		oper_chwidth = CONF_OPER_CHWIDTH_USE_HT;
 		break;
 	default:
 		/* not valid VHT bandwidth or not in CSA */
 		return eid;
 	}
 
+#ifdef CONFIG_IEEE80211BE
+	if (punct_bitmap) {
+		punct_update_legacy_bw(punct_bitmap,
+				       hapd->cs_freq_params.channel,
+				       &oper_chwidth, &chan1, &chan2);
+	}
+#endif /* CONFIG_IEEE80211BE */
+
 	*eid++ = WLAN_EID_WIDE_BW_CHSWITCH;
 	*eid++ = 3; /* Length of Wide Bandwidth Channel Switch element */
 	*eid++ = bw; /* New Channel Width */
-- 
2.45.2

